File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    void reverseRange(string &s, int l, int r) {
        while (l < r) {
            swap(s[l], s[r]);
            l++;
            r--;
        }
    }
    
    string reverseWords(string s) {
        int n = s.size(), i = 0, j = 0;
        while (i < n) {
            while (i < n && s[i] == ' ') i++;
            while (i < n && s[i] != ' ') s[j++] = s[i++];
            while (i < n && s[i] == ' ') i++;
            if (i < n) s[j++] = ' ';
        }
        s.resize(j);
        reverseRange(s, 0, s.size() - 1);
        i = 0;
        while (i < s.size()) {
            int start = i;
            while (i < s.size() && s[i] != ' ') i++;
            reverseRange(s, start, i - 1);
            i++;
        }
        return s;
    }
};

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.py
class Solution:
    def reverseWords(self, s: str) -> str:
        def reversee(brr,left,right):
            while left<=right:
                brr[left],brr[right]=brr[right],brr[left]
                left+=1
                right-=1
        brr=list(s.strip())
        reversee(brr,0,len(brr)-1)
        i=k=0
        res=[]
        while k<len(brr):
            if brr[k]!=" ":
                k+=1
            else:
                reversee(brr,i,k-1)
                res.append("".join(brr[i:k]))
                while k<len(brr) and brr[k]==" ":
                    k+=1
                i=k
        reversee(brr,i,len(brr)-1)
        res.append("".join(brr[i:]))
        return " ".join(res)
